reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/AndroidTracerTest.kt

```diff
@@ -0,0 +1,479 @@
+import com.datadog.android.tests.ktx.getString
+import com.datadog.android.trace.AndroidTracer
+import com.datadog.android.trace.Trace
+import com.datadog.android.trace.TraceConfiguration
+import com.datadog.android.trace.integration.tests.elmyr.TraceIntegrationForgeConfigurator
+import com.datadog.tools.unit.extensions.TestConfigurationExtension
+import com.datadog.tools.unit.setStaticValue
+import com.google.gson.JsonObject
+import com.google.gson.JsonParser
+import fr.xgouchet.elmyr.Forge
+import fr.xgouchet.elmyr.annotation.DoubleForgery
+import fr.xgouchet.elmyr.annotation.StringForgery
+import fr.xgouchet.elmyr.junit5.ForgeConfiguration
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import io.opentracing.util.GlobalTracer
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.offset
+import org.assertj.core.data.Offset
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.RepeatedTest
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.Extensions
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.quality.Strictness
+import java.util.concurrent.TimeUnit
+import kotlin.system.measureNanoTime
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class),
+    ExtendWith(TestConfigurationExtension::class)
+)
+@ForgeConfiguration(TraceIntegrationForgeConfigurator::class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+class AndroidTracerTest {
+
+    private lateinit var stubSdkCore: StubSDKCore
+
+    @BeforeEach
+    fun `set up`(forge: Forge) {
+        stubSdkCore = StubSDKCore(forge)
+
+        val fakeTraceConfiguration = TraceConfiguration.Builder().build()
+        Trace.enable(fakeTraceConfiguration, stubSdkCore)
+    }
+
+    @AfterEach
+    fun `tear down`() {
+        GlobalTracer::class.java.setStaticValue("isRegistered", false)
+    }
+
+    @RepeatedTest(16)
+    fun `M send trace with custom service W setService() + buildSpan() + start() + finish()`(
+        @StringForgery fakeOperation: String,
+        @StringForgery fakeService: String
+    ) {
+        // Given
+        val testedTracer = AndroidTracer.Builder(stubSdkCore).setService(fakeService).build()
+
+        // When
+        var traceId: String
+        var spanId: String
+        val fullDuration = measureNanoTime {
+            val span = testedTracer.buildSpan(fakeOperation).start()
+            traceId = span.traceIdAsHexString()
+            spanId = span.spanIdAsHexString()
+            Thread.sleep(OP_DURATION_MS)
+            span.finish()
+        }
+
+        // Then
+        val eventsWritten = stubSdkCore.eventsWritten(Feature.TRACING_FEATURE_NAME)
+        assertThat(eventsWritten).hasSize(1)
+        val event0 = JsonParser.parseString(eventsWritten[0].eventData) as JsonObject
+        println(event0)
```

Comment:
some debugging leftover here and in the tests below

Reply:
👍